File: bundler.php
<?php
/*
 * Copyright 2016 The Anonymous Press, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

if(isset($argv[1])) $in=$argv[1]; //For testing.

require_once('import_validated_orders.php'); //Generate input file from validated orders using 3dcart API.

if(!isset($in)) $in = 'input.csv';
if(!isset($out)) $out = 'incomingorders.csv';

echo "Input file: $in\r\n";

require_once('config.php');

$RemoveColumns = array('weight','unitprice','oi.weight','unitcost','itemamount','itemname','itemShipCost');
											//All other SKUs will be passed through without modification

$ShippingColumn = 'ShipMethod';
$ShippingMethods = array(					//Values containing matches for the regular expressions on the left will be replaced with the values on the right. Others will be passed through.
	'/Priority/i' => 'Priority Mail',
	'/Media/i' => 'Media Mail'
);

$path = dirname(__FILE__).'/';

if(!file_exists($path.$in)){
	die("No input file was sent to the server by 3dcart. This may indicate that no orders are marked as new, or that something is broken on the 3dcart end.");
}

$InputFile = fopen($path.$in,'r');
$OutputFile = fopen($path.$out,'w');


$header = fgetcsv($InputFile); //Get CSV header


foreach($RemoveColumns as &$name){
	$name = array_search($name,$header); //Find column numbers for columns to remove
	if($name!==FALSE) unset($header[$name]); //Clear unneeded cells from header if the name matched a name in the header
}
$header = array_values($header); //Remove empty cells from header

$SKUColumn = array_search('itemid',$header); //Find column numbers for SKU and quantity columns
$QtyColumn = array_search('numitems',$header);
$ShippingColumn = array_search($ShippingColumn,$header); //Find column number for shipping method
$OrderColumn = array_search('OrderNumber',$header);
$CouponColumn = array_search('Coupon',$header);
$ZipColumn = array_search('Zip',$header); //Zip column for ZIP+4 shim

function removecol($row,$col){
	if($col!==FALSE) unset($row[$col]);
	return array_values($row);
}

fputcsv($OutputFile,removecol($header,$CouponColumn)); //Copy CSV header to output file

$InputRows = array();
while($row = fgetcsv($InputFile)) $InputRows[]=$row; //Copy CSV rows into $InputRows[]

$CurrentOrder = ""; $CurrentSKU = ""; //Starting values
foreach($InputRows as $rownum=>&$row){ //Consolidate rows with same order number and SKU

	foreach($RemoveColumns as $num){if($num!==FALSE) unset($row[$num]);} //Clear unneeded cells from row if the name matched a name in the header
	$row=array_values($row); //Remove unneeded cells from row

	if($row[$OrderColumn] === $CurrentOrder && $row[$SKUColumn] === $CurrentSKU){ //If current row matches previous row
		$target = $rownum;
		$row[$OrderColumn] = "DELETE"; //Mark as redundant
		while(($InputRows[$target][$OrderColumn] !== $CurrentOrder || $InputRows[$target][$SKUColumn] !== $CurrentSKU) && $target > 0) $target--; //Find target row (first row with matching information)
		if($target<0) die('PANIC');
		$InputRows[$target][$QtyColumn] += $row[$QtyColumn]; //Add quantity to target row
	}
	else{
		$CurrentOrder = $row[$OrderColumn];
		$CurrentSKU = $row[$SKUColumn];
	}
}
$InputRows = array_values($InputRows);

foreach(array_keys($InputRows) as $key){ //Remove rows marked for deletion
	if($InputRows[$key][$OrderColumn] === "DELETE") unset($InputRows[$key]);
}
$InputRows = array_values($InputRows);


foreach($InputRows as &$row){ //Iterate through rows

	$sku = $row[$SKUColumn]; //Single item SKU
	$ItemQty = $row[$QtyColumn]; //Remaining quantity not yet output to output file

	if(array_search($sku,$IgnoreSKUs)!==FALSE) continue; //Skip rows containing ignored SKUs
	if(preg_match('/^ebay$/i',$row[$CouponColumn])) continue; //Skip rows for ebay orders

	foreach($ShippingMethods as $from=>$to) { //Iterate through possible shipping methods
		if(preg_match($from,$row[$ShippingColumn])){ //If shipping method matches
			$row[$ShippingColumn] = $to; //replace shipping method
			break;
		}
	}

	//ZIP+4 shim
	str_replace("-","",$row[$ZipColumn]);

	if(!isset($BundleSKUs[$sku])){ //If no bundle SKUs are available
		fputcsv($OutputFile,removecol($row,$CouponColumn)); //Copy row to output file unmodified
		continue;                  //to next row
	}

	//Below code will only be executed for input rows with SKUs for which bundle SKUs are available

	$OutputSKUs = $BundleSKUs[$sku]; //get SKUs for bundles
	$OutputSKUs[1] = $sku; //add SKU for single items

	krsort($OutputSKUs, SORT_NUMERIC); //largest bundles first
	foreach($OutputSKUs as $SKUQty=>$OutputSKU) { //Each output SKU: $SKUQty = quantity in each bundle, $OutputSKU = SKU for bundle
		if($ItemQty >= $SKUQty){ //If order is larger than bundle
			$OutputRow = $row; //Make a copy of the input row
			$OutputRow[$SKUColumn] = $OutputSKU; //Change SKU in output row to bundle SKU
			$NumBundles = floor($ItemQty/$SKUQty); //Divide remaining items by bundle quantity and round down
			$OutputRow[$QtyColumn] = $NumBundles;
			fputcsv($OutputFile,removecol($OutputRow,$CouponColumn)); //Copy modified row to output file
			$ItemQty -= $SKUQty*$NumBundles; //Subtract accounted-for items from remaining total
		}
	}
}

File: bundler.py
# Copyright 2022 Google LLC
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
import csv
import ftplib
import os.path
import re
import unicodedata
from datetime import datetime
from typing import List, Dict, Tuple

from config import Config
from opencart_db import Database, OrderStatus, OrderInfo, Items

CsvRow = Dict[str, str]


class Bundler:
    def __init__(self, db_host, db_user, db_pass, db_name, ftp_host, ftp_user, ftp_pass, ftp_path,
                 bundles, shipping, filename):
        self.db = Database(db_host, db_user, db_pass, db_name)
        self.ftp = ftplib.FTP(ftp_host, ftp_user, ftp_pass)
        self.ftp.cwd(ftp_path)
        self.filename = filename
        self.bundles = bundles
        self.shipping = shipping

    def __del__(self):
        self.ftp.quit()

    def get_orders(self) -> List[OrderInfo]:
        return self.db.get_orders_with_status(OrderStatus.VALIDATED, OrderStatus.VALIDATED_UNPAID)

    def bundle_order_items(self, items: Items) -> Items:
        bundled = {}
        for sku, qty in items.items():
            if sku in Config.ignore:
                continue
            for bundle_qty, bundle_sku in self.bundles.get(sku, {}).items():
                bundles_to_add = qty // bundle_qty
                if bundles_to_add > 0:
                    bundled[bundle_sku] = bundles_to_add
                    qty -= bundle_qty * bundles_to_add
            if qty > 0:
                bundled[sku] = qty
        return bundled

    def __map_order(self, order: OrderInfo) -> CsvRow:
        return {
            'OrderNumber': order['order_id'],
            'ShipMethod':  self.__map_ship(order['shipping_method']),
            'Comments':    '',
            'FirstName':   self.__normalize(order['shipping_firstname']),
            'LastName':    self.__normalize(order['shipping_lastname']),
            'Company':     self.__normalize(order['shipping_company']),
            'Address1':    self.__normalize(order['shipping_address_1']),
            'Address2':    self.__normalize(order['shipping_address_2']),
            'City':        self.__normalize(order['shipping_city']),
            'State':       self.__normalize(order['shipping_state']),
            'Zip':         self.__normalize(order['shipping_postcode']).replace('-', ''),
            'Phone':       self.__normalize(order['telephone']),
            'Email':       self.__normalize(order['email']),
        }

    # WarePak breaks if it receives strings with non-ASCII characters or length > 24. Woo hoo
    def __normalize(self, input: str) -> str:
        return unicodedata.normalize('NFKD', input).encode('ASCII', 'ignore').decode()[:24]

    def __map_ship(self, method: str) -> str:
        for lhs, rhs in self.shipping.items():
            if re.match('.*' + lhs, method):
                return rhs
        if method.startswith('HIDDEN'):
            return method[7:].lstrip('-').lstrip()
        return method

    FIELDS = ['OrderNumber', 'ShipMethod', 'Comments', 'FirstName', 'LastName', 'Company',
              'Address1', 'Address2', 'City', 'State', 'Zip', 'Phone', 'Email', 'itemid',
              'numitems']

    def write_csv(self, orders: List[OrderInfo]) -> Tuple[List[OrderInfo], int]:
        with open(self.filename, 'w', newline='') as output:
            writer = csv.DictWriter(output, fieldnames=self.FIELDS, quoting=csv.QUOTE_ALL,
                                    lineterminator='\n')
            writer.writeheader()
            row_count = 0
            processed_orders: List[OrderInfo] = []
            for order in orders:
                try:
                    order_info = self.__map_order(order)
                    items = self.bundle_order_items(self.db.get_order_contents(order['order_id']))
                    if Database.get_order_status(order) == OrderStatus.VALIDATED:
                        Database.set_order_status(order, OrderStatus.COMPLETE if len(
                            items) == 0 else OrderStatus.PROCESSED)
                    elif Database.get_order_status(order) == OrderStatus.VALIDATED_UNPAID:
                        Database.set_order_status(order, OrderStatus.PROCESSED_UNPAID)
                    for sku, qty in items.items():
                        writer.writerow({**order_info, 'itemid': sku, 'numitems': qty})
                        row_count += 1
                    processed_orders.append(order)
                except Exception as e:
                    print(e)
                    continue
            return processed_orders, row_count

    def update_orders(self, orders: List[OrderInfo]) -> None:
        for order in orders:
            self.db.update_order(order)

    def upload_csv(self) -> None:
        with open(self.filename, 'rb') as file:
            self.ftp.storbinary(f'STOR {os.path.basename(self.filename)}', file)


def main():
    b = Bundler(Config.db_host, Config.db_user, Config.db_pass, Config.db_name,
                Config.ftp_host, Config.ftp_user, Config.ftp_pass, Config.ftp_incoming,
                Config.bundles, Config.shipping,
                filename=datetime.now().strftime('uploaded/%Y-%m-%d_%H%M%z.csv'))
    orders, row_count = b.write_csv(b.get_orders())
    if row_count < 2:
        # Files containing no rows are pointless, and files containing one row don't get
        # processed for reasons we haven't yet been able to fathom.
        return
    b.upload_csv()
    b.update_orders(orders)


if __name__ == "__main__":
    main()
